File: javascript/quickSort.js
/**
 * Swaps two items in an array.
 * @param {Number[]} arr The array where the swap will occur.
 * @param {Number} firstIndex The first index to be used in swap process.
 * @param {Number} secondIndex The second index to be used in swap process.
 */
const swap = (arr, firstIndex, secondIndex) => {
    const temp = arr[firstIndex];
    arr[firstIndex] = arr[secondIndex];
    arr[secondIndex] = temp;
}

/**
 * Partitions the given array and returns the index of the pivot.
 * @param {Number[]} arr The array to be partitioned.
 * @param {Number} left The start point of the partitioning.
 * @param {Number} right The end point of the partitioning.
 * @returns {Number} index of the pivot
 */
const partition = (arr, left, right) => {
    const pivot = arr[right];
    let pivotPosition = left;
    for (let i = left; i < right; i++) {
        if (arr[i] <= pivot) {
            swap(arr, pivotPosition, i);
            pivotPosition++;
        }
    }
    swap(arr, pivotPosition, right);
    return pivotPosition;
}

/**
 * Recursively find the pivot and sort the given array, dividing each array as it recurses.
 * @param {Number[]} arr The array to be sorted.
 * @param {Number} left The start point of the sorting.
 * @param {Number} right The end point of the sorting.
 */
const quickSort = (arr, left, right) => {
    if (typeof left === 'undefined' && typeof right === 'undefined') {
        quickSort(arr, 0, arr.length - 1);
    } else {
        if (left < right) {
            const pivot = partition(arr, left, right);
            quickSort(arr, left, pivot - 1);
            quickSort(arr, pivot + 1, right);
        }
    }
}

exports.swap = swap;
exports.partition = partition;
exports.quickSort = quickSort;

File: csharp/DataStructuresAndAlgorithms/DataStructuresAndAlgorithms/QuickSort.cs
namespace DataStructuresAndAlgorithms
{
    public class QuickSort
    {
        /// <summary>
        /// Sort the given array by quicksort.    
        /// </summary>
        /// <param name="arr">The array to be sorted</param>
        public static void Sort(int[] arr)
        {
            Sort(arr, 0, arr.Length - 1);
        }
        /// <summary>
        /// Recursively find the pivot and sort the given array, dividing each array as it recurses.
        /// </summary>
        /// <param name="arr">The array to be sorted.</param>
        /// <param name="left">The starting point of the sorting.</param>
        /// <param name="right">The end point of the sorting.</param>
        static void Sort(int[] arr, int left, int right)
        {
            if (left < right)
            {
                int pivot = Partition(arr, left, right);
                Sort(arr, left, pivot - 1);
                Sort(arr, pivot + 1, right);
            }
        }
        /// <summary>
        /// Partitions the given array and returns the index of the pivot.
        /// </summary>
        /// <param name="arr">The array to be partitioned.</param>
        /// <param name="left">The starting point of the partitioning.</param>
        /// <param name="right">The end point of the partitioning.</param>
        /// <returns>index of the pivot</returns>
        static int Partition(int[] arr, int left, int right)
        {
            int pivotPosition = left;
            int pivot = arr[right];
            for (int i = left; i < right; i++)
            {
                if (arr[i] <= pivot)
                {
                    Swap(arr, pivotPosition, i);
                    pivotPosition++;
                }
            }
            Swap(arr, pivotPosition, right);

            return pivotPosition;
        }
        /// <summary>
        ///Swaps two items in an array
        /// </summary>
        /// <param name="arr">The array where the swap will occur.</param>
        /// <param name="firstIndex">The first index to be used in swap process.</param>
        /// <param name="secondIndex">The second index to be used in swap process.</param>
        static void Swap(int[] arr, int firstIndex, int secondIndex)
        {
            int temp = arr[firstIndex];
            arr[firstIndex] = arr[secondIndex];
            arr[secondIndex] = temp;
        }
    }
}
